custom/dhis2/models.py

```diff
@@ -0,0 +1,575 @@
+            self._fetch_tracked_entity_attributes()
+        # Set instance data keys to attribute IDs
+        inst = {self.get_te_attr_id(k): v for k, v in data.iteritems()}
+        result = self._request.put('trackedEntityInstances/' + inst['id'], inst)
+        return result
+
+    def get_top_org_unit(self):
+        """
+        Return the top-most organisation unit.
+        """
+        if settings.DHIS2_ORG_UNIT:
+            # A top organisation unit has been specified in the settings. Use that
+            __, json = self._request.get('organisationUnits',
+                                         params={'links': 'false', 'query': settings.DHIS2_ORG_UNIT})
+            return json['organisationUnits'][0]
+        # Traverse up the tree of organisation units
+        __, org_units_json = self._request.get('organisationUnits', params={'links': 'false'})
+        org_unit = org_units_json['organisationUnits'][0]
+        # The List response doesn't include parent (even if you ask for it :-| ). Request org_unit details.
+        __, org_unit = self._request.get('organisationUnits/' + org_unit['id'])
+        while True:
+            if not org_unit.get('parent'):
+                # The organisation unit with no parent is the top-most organisation unit
+                break
+            __, org_unit = self._request.get('organisationUnits/' + org_unit['parent']['id'])
+        return org_unit
+
+    def get_resource_id(self, resource, name):
+        """
+        Returns the ID of the given resource type with the given name
+        """
+        __, json = self._request.get(resource, params={'links': 'false', 'query': name})
+        if not json[resource]:
+            return None
+        if len(json[resource]) > 1:
+            raise Dhis2ApiQueryError('Query returned multiple results')
+        return json[resource][0]['id']
+
+    def get_program_id(self, name):
+        """
+        Returns the ID of the given program
+        """
+        return self.get_resource_id('programs', name)
+
+    def get_te_id(self, name):
+        """
+        Returns the ID of the given tracked entity type
+        """
+        return self.get_resource_id('trackedEntities', name)
+
+    def get_te_attr_id(self, name):
+        """
+        Returns the ID of the given tracked entity attribute
+        """
+        if name not in self._tracked_entity_attributes:
+            # Note: self.get_resource_id returns None if name not found
+            self._tracked_entity_attributes[name] = self.get_resource_id('trackedEntityAttributes', name)
+        return self._tracked_entity_attributes[name]
+
+    def gen_instances_with_unset(self, te_name, attr_name):
+        """
+        Returns a list of tracked entity instances with the given attribute name unset
+        """
+        top_ou = self.get_top_org_unit()
+        te_id = self.get_te_id(te_name)
+        attr_id = self.get_te_attr_id(attr_name)
+        page = 1
+        while True:
```

Comment:
Might be worth jumping on the DHIS2 mailing list and asking if there isn't a way to do this in a query: https://launchpad.net/~dhis2-users